flink-runtime/src/test/java/org/apache/flink/runtime/jobmanager/DefaultJobGraphStoreTest.java

```diff
@@ -0,0 +1,306 @@
+			})
+			.build();
+
+		final JobGraphStore jobGraphStore = createAndStartJobGraphStore(stateHandleStore);
+		jobGraphStore.putJobGraph(testingJobGraph);
+
+		final JobGraph actual = addFuture.get(timeout, TimeUnit.MILLISECONDS);
+		assertThat(actual.getJobID(), is(testingJobGraph.getJobID()));
+	}
+
+	@Test
+	public void testPutJobGraphWhenAlreadyExist() throws Exception {
+		final CompletableFuture<Tuple3<String, String, JobGraph>> replaceFuture = new CompletableFuture<>();
+		final String resourceVersion = "100";
+		final AtomicBoolean alreadyExist = new AtomicBoolean(false);
+		final TestingStateHandleStore<JobGraph> stateHandleStore = builder
+			.setExistsFunction(ignore -> {
+				if (alreadyExist.get()) {
+					return resourceVersion;
+				} else {
+					alreadyExist.set(true);
+					return NON_EXIST_RESOURCE_VERSION;
+				}
+			})
+			.setAddFunction((ignore, state) -> jobGraphStorageHelper.store(state))
+			.setReplaceConsumer(replaceFuture::complete)
+			.build();
+
+		final JobGraphStore jobGraphStore = createAndStartJobGraphStore(stateHandleStore);
+		jobGraphStore.putJobGraph(testingJobGraph);
+		// Replace
+		jobGraphStore.putJobGraph(testingJobGraph);
+
+		final Tuple3<String, String, JobGraph> actual = replaceFuture.get(timeout, TimeUnit.MILLISECONDS);
+		assertThat(actual.f0, is(testingJobGraph.getJobID().toString()));
+		assertThat(actual.f1, is(resourceVersion));
+		assertThat(actual.f2.getJobID(), is(testingJobGraph.getJobID()));
+	}
+
+	@Test
+	public void testRemoveJobGraph() throws Exception {
+		final CompletableFuture<JobID> removeFuture = new CompletableFuture<>();
+		final TestingStateHandleStore<JobGraph> stateHandleStore = builder
+			.setAddFunction((ignore, state) -> jobGraphStorageHelper.store(state))
+			.setRemoveFunction(name -> removeFuture.complete(JobID.fromHexString(name)))
+			.build();
+
+		final JobGraphStore jobGraphStore = createAndStartJobGraphStore(stateHandleStore);
+		jobGraphStore.removeJobGraph(testingJobGraph.getJobID());
+
+		jobGraphStore.putJobGraph(testingJobGraph);
+		jobGraphStore.removeJobGraph(testingJobGraph.getJobID());
+		final JobID actual = removeFuture.get(timeout, TimeUnit.MILLISECONDS);
+		assertThat(actual, is(testingJobGraph.getJobID()));
+	}
+
+	@Test
+	public void testRemoveJobGraphWithNonExistName() throws Exception {
+		final CompletableFuture<JobID> removeFuture = new CompletableFuture<>();
+		final TestingStateHandleStore<JobGraph> stateHandleStore = builder
+			.setRemoveFunction(name -> removeFuture.complete(JobID.fromHexString(name)))
+			.build();
+
+		final JobGraphStore jobGraphStore = createAndStartJobGraphStore(stateHandleStore);
+		jobGraphStore.removeJobGraph(testingJobGraph.getJobID());
+
+		try {
+			removeFuture.get(timeout, TimeUnit.MILLISECONDS);
+			fail("We should get an expected timeout because we are removing a non-existed job graph.");
+		} catch (TimeoutException ex) {
+			// expected
+		}
+		assertThat(removeFuture.isDone(), is(false));
```

Comment:
```
@Test(expected = TimeoutException.class)
...
try {
	removeFuture.get(timeout, TimeUnit.MILLISECONDS);
} finally {
	assertThat(removeFuture.isDone(), is(false));
}
```

Reply:
Hmm. Actually, I want to fail the test with some descriptive reason.